crates/core_simd/examples/dot_product.rs

```diff
@@ -0,0 +1,54 @@
+// Code taken from the `packed_simd` crate
+// Run this code with `cargo test --example dot_product`
+#![feature(array_chunks)]
+use core_simd::*;
+
+/// This is your barebones dot product implementation: 
+/// Take 2 vectors, multiply them element wise and *then*
+/// add up the result. In the next example we will see if there
+///  is any difference to adding as we go along multiplying.
+pub fn dot_prod_0(a: &[f32], b: &[f32]) -> f32 {
+    assert_eq!(a.len(), b.len());
+
+    a.iter()
+    .zip(b.iter())
+    .map(|a, b| a * b)
+    .sum()
+}
+
+pub fn dot_prod_1(a: &[f32], b: &[f32]) -> f32 {
+    assert_eq!(a.len(), b.len());
+    a.iter()
+    .zip(b.iter())
+    .fold(0.0, |a, b| a * b)
+}
+
+pub fn dot_prod_simd_0(a: &[f32], b: &[f32]) -> f32 {
+    assert_eq!(a.len(), b.len());
+
+    // TODO handle remainder when a.len() % 4 != 0
```

Comment:
I've put together an example that handles remainder, if you'd like to use it: <https://rust.godbolt.org/z/eMPEnc9or>

```rust
const LANES: usize = 8;
pub fn dot_prod_simd(a: &[f32], b: &[f32]) -> f32 {
    assert_eq!(a.len(), b.len());

    let (a_extra, a_chunks) = a.as_rchunks();
    let (b_extra, b_chunks) = b.as_rchunks();

    // These are always true, but for emphasis:
    assert_eq!(a_chunks.len(), b_chunks.len());
    assert_eq!(a_extra.len(), b_extra.len());

    let mut sums = [0.0; LANES];
    for ((x, y), d) in zip(a_extra, b_extra).zip(&mut sums) {
        *d = x * y;
    }

    let mut sums = Simd::from_array(sums);
    zip(a_chunks, b_chunks)
        .for_each(|(x, y)| {
            sums += Simd::from_array(*x) * Simd::from_array(*y);
        });

    sums.reduce_sum()
}
```

Reply:
Awesome! I'll add it to the batch.